### PR TITLE
fix: reset streams on prerender only for affected browser versions

### PR DESCRIPTION
**Pull request by jiexi**
## **Description**

The prerender regression in chromium has been resolved already. Resetting streams on prerendered pages no longer required for newer chromium versions and is actually the cause of unresponsive provider in many cases. This PR modifies the existing workaround to target only the affected chromium ranges from >=113 to <121.

My bisection of chromium for the fix landed on this range of commits: https://chromium.googlesource.com/chromium/src/+log/55b4344edfb41dda980d197743f25a2841d498a4..c13107c16780c195bd5ec003d9198d87cdcc59dd

@shanejonas and I have verified that the prerender issue has been resolved in the latest chrome stable. I have verified that the workaround is causing an issue in develop that is no longer seen in this branch that reverts it.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/24293?quickstart=1)

## **Related issues**

See: https://github.com/MetaMask/metamask-extension/pull/19727

## **Manual testing steps**
Using chrome, of course.
Visit `chrome://settings/?search=preload` and ensure it is enabled and on "standard"

1. On develop, make a build and load it
2. Visit `https://voyager-snap.linea.build/`
3. Install the snap and connect
4. Copy `https://voyager-snap.linea.build/` to your clipboard
5. Close all instances of `https://voyager-snap.linea.build/` that may be open
6. Rapidly, open a new tab, paste the url, hit enter, see if the page is stuck on "Loading", if not close the tab and repeat
7. Eventually the page should be get prerendered and stuck in a broken state

Do the same steps above, but using the changes in this branch. The page should never be stuck with "Loading" if you are on a chromium version from 102 to 112, or 121 onwards.

### **Before**

https://github.com/MetaMask/metamask-extension/assets/918701/6bda7a5d-12d2-46e4-b5d6-8562c3143977

### **After**

https://github.com/MetaMask/metamask-extension/assets/918701/14289f0e-89eb-4534-8815-615a57e0386b

### Browser version targeting proof

The following screenshots show the return value for `getIsBrowserPrerenderBroken()` on various chromium versions

120:

Screenshot 2024-04-30 at 12 47 57 PM

119:

Screenshot 2024-04-30 at 12 50 57 PM

113:
Screenshot 2024-04-30 at 11 22 20 AM

112:
Screenshot 2024-04-30 at 11 20 43 AM

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: ui/helpers/constants/common.ts [hunk 1]
```diff
@@ -23,3 +23,24 @@ export const OUTDATED_BROWSER_VERSIONS = {
   // See https://en.wikipedia.org/wiki/History_of_the_Opera_web_browser
```
Comment (Gudahtt) [inline ui/helpers/constants/common.ts:23]: Probably we could treat Opera 99-(107?) as affected as well, but maybe not worth worrying about given the impact of the bug. We can adjust this later if we need to. Who knows what they're changing in their fork.

File: shared/modules/browser-runtime.utils.js [hunk 2]
```diff
@@ -53,3 +58,13 @@ export function checkForLastErrorAndWarn() {
 
   return error;
 }
+
+export function getIsBrowserPrerenderBroken(
```
Comment (Gudahtt) [inline shared/modules/browser-runtime.utils.js:62]: Nit: A JSDoc comment would be appreciated here!

File: ui/helpers/constants/common.ts [hunk 3]
```diff
@@ -23,3 +23,24 @@ export const OUTDATED_BROWSER_VERSIONS = {
...
   opera: '<76',
 };
+export const BROKEN_PRERENDER_BROWSER_VERSIONS = {
```
Comment (Gudahtt) [inline ui/helpers/constants/common.ts:26]: Nit: Probably worthwhile having at least brief descriptions of these constants as well

**Review by Gudahtt — DISMISSED** [review]
LGTM!

**Review by adonesky1 — DISMISSED** [review]
LGTM! Thanks for so much great digging here @jiexi!

File: shared/modules/browser-runtime.utils.js (hunk 2 above)
Comment (jiexi) [reply shared/modules/browser-runtime.utils.js:72]: good idea https://github.com/MetaMask/metamask-extension/pull/24317/commits/821221265383002f1de0cde8765d154f2a9047a6

**Review by Gudahtt — DISMISSED** [review]

File: ui/helpers/constants/common.ts (hunk 3 above)
Comment (jiexi) [reply ui/helpers/constants/common.ts:34]: agreed https://github.com/MetaMask/metamask-extension/pull/24317/commits/08279efdf8ad998044bae5cd06f1d46e99ddee2a

**Review by Gudahtt — APPROVED** [review]

**Review by adonesky1 — APPROVED** [review]

**Pull request merged**